# PROYECTO MÓDULO 1

## Optimización Cementera

## Objetivos

**CEMENTERA**

Existe una cementera que cuenta con 4 plantas(A, B, C y D), para producir unidades(costales) de 2 clases diferentes de cementos(clase 1 y 2), cada una de estas plantas cuentan con equipo, personal, materias primas y límites de producción de unidades de cemento diferentes por lo tanto la utilidad que obtienen de cada clase de cemento es diferente, además cada una tiene un limite de producción diferente.

* La planta A obtiene una utilidad por cada unidad de cemento producido de clase:
  * Clase 1 de \$50.
  * Clase 2 de \$75.
  * Y tiene un limite de producción de 1050 unidades por día.
  
* La planta B obtiene una utilidad por cada unidad de cemento producido de clase:
  * Clase 1 de \$40.
  * Clase 2 de \$95.
  * Y tiene un limite de producción de 600 unidades por día.

* La planta C obtiene una utilidad por cada unidad de cemento producido de clase:
  * Clase 1 de \$65.
  * Clase 2 de \$55.
  * Y tiene un limite de producción de 950 unidades por día.
  
* La planta D obtiene una utilidad por cada unidad de cemento producido de clase:
  * Clase 1 de \$60.
  * Clase 2 de \$65.
  * Y tiene un limite de producción de 450 unidades por día.
  
Considerando que la demanda diaria que existe de cemento es de: cemento clase 1; 1750 unidades diarias y cemento clase 2; 950 unidades diarias. 

**¿Cúantas unidades de cemento de cada clase debe producir cada planta  al día, para obtener la utilidad máxima posible?**


**Objetivo Principal**

* Obtener las ganancias máximas en la venta de cementos.

**Objetivo Específicos**

* Conocer la cantidad de cemento a fabricar de cada clase en cada planta, de acuerdo a la demanda y limitantes de producción.

## Modelo del problema

**Función a Optimizar**

$$50x_1 + 75x_2 + 40x_3 + 95x_4 + 65x_5 + 55x_6 + 60x_7 + 65x_8$$

*Buscamos la cantidad optima a producir diariamente de cada clase de cemento por cada una de las planas para obtener las máximas ganancias posibles.*

Donde:

- $x_1$ unidades de cemento clase 1 producido por la planta A.
- $x_2$ unidades de cemento clase 2 producido por la planta A.
- $x_3$ unidades de cemento clase 1 producido por la planta B.
- $x_4$ unidades de cemento clase 2 producido por la planta B.
- $x_5$ unidades de cemento clase 1 producido por la planta C.
- $x_6$ unidades de cemento clase 2 producido por la planta C.
- $x_7$ unidades de cemento clase 1 producido por la planta D.
- $x_8$ unidades de cemento clase 2 producido por la planta D.

*Definimos nuestras variables de decisión.*

**Restricciones:**

- $x_1 + x_2 \leq 1050$
- $x_3 + x_4 \leq 600$
- $x_5 + x_6 \leq 950$
- $x_7 + x_8 \leq 450$
- $x_1 + x_3 + x_5 + x_7 \leq 1750$
- $x_2 + x_4 + x_6 + x_8 \leq 950$

*Establecemos nuestras restricciones dadas por el limite de producción de cada planta y la demanda de cada clase de cemento.*

## Solución del problema

In [1]:
# Importamos todas las librerías que necesitaremos
import numpy as np
import scipy.optimize as opt
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
# Crearmos las mátrices para resolver el problema
c = np.array([-50,-75,-40,-95,-65,-55,-60,-65])

A = np.array([[1,1,0,0,0,0,0,0],[0,0,1,1,0,0,0,0],[0,0,0,0,1,1,0,0],[0,0,0,0,0,0,1,1],[1,0,1,0,1,0,1,0],[0,1,0,1,0,1,0,1]])

b = np.array([1050,600,950,450,1750,950])

In [3]:
# Resolvemos el problema con la función linprog
solucion = opt.linprog(c,A,b)

In [4]:
solucion

     con: array([], dtype=float64)
     fun: -189500.0
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([350.,   0.,   0.,   0.,   0.,   0.])
  status: 0
 success: True
       x: array([350., 350.,   0., 600., 950.,   0., 450.,   0.])

In [5]:
solucion.x

array([350., 350.,   0., 600., 950.,   0., 450.,   0.])

In [6]:
import pandas as pd

In [12]:
file_name = 'proyecto_cementera.csv'
data = pd.read_csv(file_name)
data

Planta       A      B      C      D  \
0                         Clase 1     $50    $40    $65    $60   
1                         Clase 2     $75    $95    $55    $65   
2  Limite de producción(unidades)  1050 u  600 u  950 u  450 u   

   Demanda(unidades)  
0             1750 u  
1              950 u  
2                  .

### Conclusión

*Ahora podemos concluir que la producción optima de cada clase de cemento por parte de cada plante es de:*

- Planta A Clase 1: 350 unidades.
- Planta A Clase 2: 350 unidades.


- Planta B Clase 1: 0   unidades.
- Planta B Clase 2: 600 unidades.


- Planta C Clase 1: 950 unidades.
- Planta C Clase 2: 0   unidades.


- Planta D Clase 1: 450 unidades.
- Planta D Clase 2: 0   unidades.


*Y a su vez esto cumple con las restricciones de producción*

- Total de unidades producidas Planta A :  700 unidades.
- Limite de unidades a producir Planta A: 1050 unidades.


- Total de unidades producidas Planta B : 600 unidades.
- Limite de unidades a producir Planta B: 600 unidades.


- Total de unidades producidas Planta C : 950 unidades.
- Limite de unidades a producir Planta C: 950 unidades.


- Total de unidades producidas Planta D:  450 unidades.
- Limite de unidades a producir Planta D: 450 unidades.


*Y también con la restricción de demanda*

- Total de cemento clase 1 producido por todas las plantas diariamente: 1720 unidades.
- Limite de cemento clase 1 a producir diariamente                    : 1750 unidades.


- Total de cemento clase 2 producido por todas las plantas diariamente: 950  unidades.
- Limite de cemento clase 2 a producir diariamente                    : 950  unidades.

*Finalmente con toda esta producción obtuvimos una utilidad total diaria de:*

- **$189,500**